packages/web3torrent/src/library/web3torrent-lib.ts

```diff
@@ -297,24 +302,34 @@ export default class WebTorrentPaidStreamingClient extends WebTorrent {
           break;
         case PaidStreamingExtensionNotices.MESSAGE:
           message = JSON.parse(data.message);
-          await this.paymentChannelClient.pushMessage(message);
-          channelId = message.data.channelId;
-          channelState = this.paymentChannelClient.channelCache[channelId];
```

Comment:
I think this was probably the source of the "not your turn" errors we were seeing. We were pulling information out of the `channelCache`, and this was only being updated asynchronously (I receive the message and push it into my wallet, and this promise resolves; some time later, a channel updated event fires and the `channelCache` is updated. It's not clear how to await that second event here, as we would need to if we wanted to avoid the race condition).

The solution here is to grab the state straight out of the received message. Downside is we don't do any kind of signature of balances (`validTransition`) checks on that state.

Previously, we updated the `channelCache` on the way "in", which is another way of avoiding the problem.

Reply:
In the end I moved this code into an `onChannelUpdated` callback, which is definitely the best solution.